### PR TITLE
Adding mdbook support/CI for PRs

### LINKED ISSUE #6: Make the book accessible through GitHub Pages
Our initial CI script in copied from the rustc dev guide, but we don't currently have a `gh-pages` branch, so will probably need to do some more work to get it readable.

### PR DESCRIPTION
**Pull request by simonsan**
Fixes #6 

Output: https://simonsan.github.io/std-dev-guide/

@KodrAus Feedback welcome :)

### CODE REVIEW

**Comment by camelid** [comment]
Is this based on equivalent code for the rustc-dev-guide? Ideally we'd keep the infra code in sync somehow...

**Comment by KodrAus** [comment]
I think the rustc dev guide is using Travis CI for its mdbook support. Was there a reason not to use Actions on that side too?

**Comment by camelid** [comment]
> Was there a reason not to use Actions on that side too?

Not sure, but I think we're going to try to switch to Actions. And new stuff is being put on Actions: https://github.com/rust-lang/rustc-dev-guide/pull/1037

**Comment by JohnTitor** [comment]
IIRC third-party actions shouldn't be used generally for security reasons (I think @pietroalbini has some thoughts about it).

**Comment by KodrAus** [comment]
> IIRC third-party actions shouldn't be used generally for security reasons

Is this just for third-party actions that we pass secrets into? Or can any third-party action access secrets without us explicitly passing them through? I'd be interested to hear more about that.

**Comment by JohnTitor** [comment]
Given https://github.com/rust-lang/crates.io/pull/2951#issuecomment-720568541 I think we should avoid using them regardless of secrets.

**Comment by KodrAus** [comment]
Thanks @JohnTitor, that seems fair to me.

@simonsan Maybe we could look at what it would take to inline the pieces of these actions we actually use? If we want to move the Rustc guide to Actions too at some point it seems like we may as well try get something working with it here too, then look at making things consistent.

**Comment by camelid** [comment]
For the "setup mdBook" part I would think you could just copy the commands that are run in Travis from the rustc-dev-guide since they're just shell commands. Not sure about the deploy; you might want to look through the `.travis.yml` file and the `ci/` directory in `rust-lang/rustc-dev-guide`.

**Comment by KodrAus** [comment]
I imagine the deploy would be the usual dance of creating an empty Git repository, copying the HTML content into it and force pushing that to the `gh-pages` branch.

**Comment by camelid** [comment]
I think the rustc-dev-guide actually just force-pushes to a [branch called `gh-pages`](https://github.com/rust-lang/rustc-dev-guide/tree/gh-pages).

**Comment by simonsan** [comment]
These scripts in `./ci` there is a markdown lint action, we use in `patterns`:

```
  markdown-lint:
    runs-on: ubuntu-20.04
    steps:
      - uses: actions/checkout@v2
      - name: Lint all files recursively
        uses: avto-dev/markdown-lint@v1
        with:
          config: '.markdownlint.yaml' 
          args: '**/*.md'
```

Would you also not consider using this and instead use the script? MDL is easy to setup with a `yaml` file:
https://github.com/rust-unofficial/patterns/blob/master/.markdownlint.yaml

Same for markdown link check:
```
  markdown-link-check:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master
    - uses: gaurav-nelson/github-action-markdown-link-check@v1
      with:
        use-verbose-mode: 'yes'
        use-quiet-mode: 'yes'
        config-file: '.github/workflows/url-check-config.json'
        check-modified-files-only: 'yes'
 ```
 
 These actions shouldn't modify anything and push something to branches and mostly run during a PR, so I'm not sure whether there should be any security concerns about it, while I understand these for mdbook actions and deployment.

File: .env [hunk 1]
```diff
@@ -0,0 +1,7 @@
+MDBOOK_VERSION=0.4.6
+MDBOOK_LINKCHECK_VERSION=0.7.2
+MDBOOK_TOC_VERSION=0.6.1
+
+GIT_DEPLOY_DIR=book   
+GIT_DEPLOY_BRANCH=gh-pages
+GIT_DEPLOY_REPO=https://github.com/simonsan/std-dev-guide
```
Comment (simonsan) [inline .env:7]: Will change that, before merging, just for now for testing purpose for the output on https://simonsan.github.io/std-dev-guide/

File: .github/workflows/gh-pages.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+name: github pages
+
+on:
+  push:
+    branches:
+      - master
+      - mdbook
```
Comment (simonsan) [inline .github/workflows/gh-pages.yml:7]: Same here, will remove that before mergin

**Review by KodrAus — APPROVED** [review]
Thanks for working on this @simonsan! Once we update the variables this looks good to me.

**Comment by simonsan** [comment]
@KodrAus Should I remove the `.travis.yml`?

**Comment by simonsan** [comment]
@KodrAus I've added mdbook build/test to test the building process of the book and the code samples during PR. Do you think that makes sense? I have prepared fixes for the files already so `mdbook test` runs through successfully that I can PR as well.

**Comment by simonsan** [comment]
Blocking:
#10 - One link is broken and one link is missing
#11 - Makes code samples tests pass for now

**Comment by KodrAus** [comment]
Thanks @simonsan! I merged in your fixes.

Removing the `.travis.yml` sounds good to me too 🙂 Save Travis a little compute.

**Comment by simonsan** [comment]
> Removing the `.travis.yml` sounds good to me too 🙂 Save Travis a little compute.

Done :)

**Comment by KodrAus** [comment]
Thanks so much for working on this @simonsan! :bow:

**Comment by simonsan** [comment]
@KodrAus Seems there is still an issue with that CNAME. Anything to fix from my side?

**Comment by KodrAus** [comment]
@simonsan It's all good, we've got #12 opened to track getting that sorted. Thanks again for working on this!

**Pull request merged**